runtime/j9vm/javanextvmi.c

```diff
@@ -32,10 +40,96 @@ JVM_InitializeFromArchive(JNIEnv *env, jclass clz)
 #endif /* JAVA_SPEC_VERSION >= 11 */
 
 #if JAVA_SPEC_VERSION >= 14
+typedef struct GetStackTraceElementUserData {
+	J9ROMClass *romClass;
+	J9ROMMethod *romMethod;
+	UDATA bytecodeOffset;
+} GetStackTraceElementUserData;
+
+static UDATA
+getStackTraceElementIterator(J9VMThread *vmThread, void *voidUserData, UDATA bytecodeOffset, J9ROMClass *romClass, J9ROMMethod *romMethod, J9UTF8 *fileName, UDATA lineNumber, J9ClassLoader *classLoader)
+{
+	GetStackTraceElementUserData *userData = voidUserData;
+
+	/* We are done, only first stack frame is needed. */
+	userData->romClass = romClass;
+	userData->romMethod = romMethod;
+	userData->bytecodeOffset = bytecodeOffset;
+
+	return FALSE;
+}
+
+#ifdef DEBUG_BCV
+static void cfdumpBytecodePrintFunction(void *userData, char *format, ...)
+{
+	PORT_ACCESS_FROM_PORT((J9PortLibrary*)userData);
+	va_list args;
+	char outputBuffer[512] = {0};
+
+	va_start(args, format);
+	j9str_vprintf(outputBuffer, 512, format, args);
+	va_end(args);
+	j9tty_printf(PORTLIB, "%s", outputBuffer);
+}
+#endif
+
 JNIEXPORT jstring JNICALL
 JVM_GetExtendedNPEMessage(JNIEnv *env, jthrowable throwableObj)
 {
-	/* Returning NULL to allow JDK14 compilation, https://github.com/eclipse/openj9/issues/7500 */
-	return NULL;
+	J9VMThread *vmThread = (J9VMThread *)env;
+	J9JavaVM *vm = vmThread->javaVM;
+	jobject msgObjectRef = NULL;
+	
+	Trc_SC_GetExtendedNPEMessage_Entry(vmThread, throwableObj);
+	if (J9_ARE_ANY_BITS_SET(vm->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_SHOW_EXTENDED_NPEMSG)) {
+		J9InternalVMFunctions const * const vmFuncs = vm->internalVMFunctions;
+		char *npeMsg = NULL;
+		GetStackTraceElementUserData userData = {0};
+		J9BytecodeVerificationData *bcvd = vm->bytecodeVerificationData;
+		
+		Trc_SC_GetExtendedNPEMessage_Entry2(vmThread, throwableObj);
+		vmFuncs->internalEnterVMFromJNI(vmThread);
+		userData.bytecodeOffset = UDATA_MAX;
+		vmFuncs->iterateStackTrace(vmThread, (j9object_t*)throwableObj, getStackTraceElementIterator, &userData, FALSE);
+		if ((NULL != userData.romClass)
+			&& (NULL != userData.romMethod)
+			&& (UDATA_MAX != userData.bytecodeOffset)
+			&& (NULL != bcvd)
+		) {
+#ifdef DEBUG_BCV
```

Comment:
Please use `#if defined` rather than `#ifdef`, and please comment the else and endif directives (for all ifdefs in this file and any others you may have added).

Reply:
Fixed.